realm/src/main/java/io/realm/dynamic/RealmObjectSchema.java

```diff
@@ -0,0 +1,491 @@
+        if (!table.hasSearchIndex(columnIndex)) {
+            throw new IllegalArgumentException("Field is not indexed: " + fieldName);
+        }
+        table.removeIndex(columnIndex);
+        return this;
+    }
+
+    /**
+     * Adds a primary key to a given field. This is the same as adding the {@code @PrimaryKey} annotation on the field.
+     *
+     * @param fieldName Field to add name to rename.
+     * @return The updated schema.
+     * @throws IllegalArgumentException if field name doesn't exists, the field cannot be a primary key or it already
+     * has a primary key defined.
+     */
+    public RealmObjectSchema addPrimaryKey(String fieldName) {
+        checkEmpty(fieldName);
+        checkFieldExists(fieldName);
+        table.setPrimaryKey(fieldName);
+        return this;
+    }
+
+    /**
+     * Removes the primary key from this class. This is the same as removing the {@code @PrimaryKey} annotation from the
+     * class.
+     *
+     * @return The updated schema.
+     * @throws IllegalArgumentException if the class doesn't have a primary key defined.
+     */
+    public RealmObjectSchema removePrimaryKey() {
+        if (!table.hasPrimaryKey()) {
+            throw new IllegalStateException(getClassName() + " doesn't have a primary key.");
+        }
+        table.setPrimaryKey("");
+        return this;
+    }
+
+    /**
+     * Creates an object with default values. Classes with a primary key defined must use {@link #createObject(Object)}
+     * instead.
+     *
+     * @return The new object. All fields will have default values for their type.
+     * @throws IllegalStateException if the class have a primary key defined.
+     */
+    public DynamicRealmObject createObject() {
+        if (table.hasPrimaryKey()) {
+            throw new IllegalStateException("Class requires a primary key value. Use createObject(primaryKeyValue) instead.");
+        }
+        long rowIndex = table.addEmptyRow();
+        return new DynamicRealmObject(realm, table.getCheckedRow(rowIndex));
+    }
+
+    /**
+     * Creates an object with a given primary key. Classes without a primary key defined must use {@link #createObject()}
+     * instead.
+     *
+     * @return The new object. All fields will have default values for their type, except for the primary key field which
+     * will have the provided value.
+     * @throws IllegalArgumentException if the primary key value is of the wrong type.
+     * @throws IllegalStateException if the class doesn't have a primary key defined.
+     */
+    public DynamicRealmObject createObject(Object primaryKeyValue) {
+        long index = table.addEmptyRowWithPrimaryKey(primaryKeyValue);
+        return new DynamicRealmObject(realm, table.getCheckedRow(index));
+    }
+
+    private void setModifiers(long columnIndex, Set<RealmModifier> modifiers) {
+        if (modifiers != null && modifiers.size() > 0) {
+            if (modifiers.contains(RealmModifier.INDEXED)) {
```

Comment:
It is possible to remove the search index. It could be done by saying that if `RealmModifier.INDEXED` isn't in `modifiers` and the field has a search index, then it is removed.